### PR TITLE
bug-45 don't bundle artifacts more than once; give precedence to instructions on non-transitive dependencies

### PR DESCRIPTION
**Pull request by lukejackson**
Test case and fix for the issue described in bug 45, which I have come across in my own projects

The change is to bundle all directly specified artifacts first, before moving on to transitive dependencies. when processing a dependency, first check whether it has already been bundled, in which case skip it. previously the dependency was rebundled, which is inefficient and can overwrite a dependency bundled with instructions with a new copy without instructions.

### CODE REVIEW

**Comment by lukejackson** [comment]
once https://github.com/reficio/p2-maven-plugin/pull/71 is merged i'll rebase this change so the build succeeds

File: src/test/integration/bug-45-it/pom.xml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                                    <id>ch.qos.logback:logback-core:1.1.2</id>
+                                    <override>true</override>
+                                    <instructions>
+                                        <Foo>Bar</Foo>
+                                    </instructions>
+                                </artifact>
+                                <!-- logback-classic depends on logback-core -->
+                                <artifact>
+                                    <id>ch.qos.logback:logback-classic:1.1.2</id>
+                                </artifact>
```
Comment (lukejackson) [inline src/test/integration/bug-45-it/pom.xml]: previously, the plugin would have rebundled logback-core as a transitive dependency of logback-classic, overwriting the previous bundle (and its custom header values)

**Comment by lukejackson** [comment]
merged master and tests are passing now

**Comment by tombujok** [comment]
Awesome! Thanks!

**Pull request merged**